File: map2SubjScale.py
# %%% map2SubjScale.m %%%
import numpy as np
import scipy.io as sio
from myMapping import myMapping


# function [OPS, TPS, IPS, APS] = map2SubjScale(qTarget, qInterf, qArtif, qGlobal)
def map2SubjScale(qTarget, qInterf, qArtif, qGlobal):
    # q = [qGlobal; qTarget; qInterf; qArtif];
    q = np.array([qGlobal, qTarget, qInterf, qArtif])

    # % Log-mapping
    # q=max(min(log((1+q)./(1-q)),5.5),-5.5);
    q_mapped = np.maximum(np.minimum(np.log((1.0 + q) / (1.0 - q)), 5.5), -5.5)

    taskQ = np.zeros(4)
    # for nTask = 1:4
    for nTask in range(4):
        # load(sprintf('paramTask%d.mat',nTask));
        # Note: Using scipy.io.loadmat to dynamically load the neural parameters
        mat_data = sio.loadmat(f"paramTask{nTask + 1}.mat")

        # Extract variables from mat structure
        # W, b, v, a and selec
        W = mat_data['W']
        b = mat_data['b']
        v = mat_data['v']
        a = mat_data['a']
        # Note: MATLAB arrays are 1-based, we must subtract 1 from the feature indices
        selec = mat_data['selec'].flatten() - 1

        # taskQ(nTask) = myMapping(q(selec),W,b,v,a);
        taskQ[nTask] = myMapping(q_mapped[selec], W, b, v, a)

    # OPS = taskQ(1);
    # TPS = taskQ(2);
    # IPS = taskQ(3);
    # APS = taskQ(4);
    OPS = taskQ[0]
    TPS = taskQ[1]
    IPS = taskQ[2]
    APS = taskQ[3]

    return OPS, TPS, IPS, APS

File: references/peass_master_22c7fc4e/v1.0/map2SubjScale.m
function [OPS, TPS, IPS, APS] = ...
    map2SubjScale(qTarget, qInterf, qArtif, qGlobal)
% Non-linear mapping to subjective scale
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Version 1.0
% Copyright 2010 Valentin Emiya (INRIA).
% This software is distributed under the terms of the GNU Public License
% version 3 (http://www.gnu.org/licenses/gpl.txt).
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

q = [qGlobal,qTarget, qInterf, qArtif];

taskQ = NaN(4,1);
for nTask = 1:4
    load(sprintf('paramTask%d.mat',nTask));
    % Check number of inputs
    switch size(W,2)
        case 1
            x = q(nTask);
        case 3
            x = q(2:4);
        case 4
            x = q;
        otherwise
            error('map2SubjScale:wrongParameters','Wrong parameters');
    end
    % Non-linear mapping
    taskQ(nTask) = myMapping(x(:),W,b,v);
end

% Threshold to output values in [0 100]
taskQ = min(100,max(0,taskQ));

OPS = taskQ(1);
TPS = taskQ(2);
IPS = taskQ(3);
APS = taskQ(4);
return